pkg/cmd/util/tokencmd/negotiate_gssapi.go

```diff
@@ -0,0 +1,211 @@
+// +build gssapi
+
+package tokencmd
+
+import (
+	"encoding/base64"
+	"errors"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/apcera/gssapi"
+	"github.com/golang/glog"
+)
+
+func NegotiateEnabled() bool {
+	return true
+}
+
+// NegotiateChallengeHandler manages a challenge negotiation session
+// it is single-host, single-use only, and not thread-safe
+type NegotiateChallengeHandler struct {
+	// handle to a loaded gssapi lib
+	lib *gssapi.Lib
+	// error seen when loading the lib
+	loadError error
+	// lock to make sure we only load it once
+	loadOnce sync.Once
+
+	// service name of the server we are negotiating against
+	name *gssapi.Name
+	// security context holding the state of the negotiation
+	ctx *gssapi.CtxId
+	// flags indicating what options we want for the negotiation
+	// TODO: surface option for mutual authentication, e.g. gssapi.GSS_C_MUTUAL_FLAG
+	flags uint32
+
+	// tracks whether the managed resources have been released
+	released bool
+	// tracks whether gssapi requires more calls to InitSecContext before considering the negotiation complete
+	continueNeeded bool
+}
+
+func NewNegotiateChallengeHandler() ChallengeHandler {
+	return &NegotiateChallengeHandler{}
+}
+
+func (c *NegotiateChallengeHandler) CanHandle(headers http.Header) bool {
+	// Make sure this is a negotiate request
+	isNegotiate, _, err := getNegotiateToken(headers)
+	if err != nil || !isNegotiate {
+		return false
+	}
+	// Make sure we can get a gssapi library
+	_, err = c.loadLib()
+	return err == nil
+}
+
+func (c *NegotiateChallengeHandler) HandleChallenge(requestURL string, headers http.Header) (http.Header, bool, error) {
+	if c.released {
+		glog.V(5).Infof("already completed negotiation, cannot start another one")
+		return nil, false, errors.New("already completed negotiation, cannot start another one")
+	}
+
+	lib, err := c.loadLib()
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Get incoming token
+	_, incomingToken, err := getNegotiateToken(headers)
+	if err != nil {
+		return nil, false, err
+	}
+	incomingTokenBuffer, err := lib.MakeBufferBytes(incomingToken)
+	if err != nil {
+		return nil, false, err
+	}
+	defer incomingTokenBuffer.Release()
+
+	if c.ctx == nil {
+		u, err := url.Parse(requestURL)
+		if err != nil {
+			return nil, false, err
+		}
+
+		glog.V(5).Infof("importing service name %s", u.Host)
+		nameBuf, err := lib.MakeBufferString("HTTP@" + u.Host)
+		if err != nil {
+			return nil, false, err
+		}
+		defer nameBuf.Release()
+
+		name, err := nameBuf.Name(lib.GSS_C_NT_HOSTBASED_SERVICE)
+		if err != nil {
+			return nil, false, err
+		}
+		c.name = name
+		c.ctx = lib.GSS_C_NO_CONTEXT
+	}
+
+	// Call InitSecContext to validate the incoming token (if given), and get our outgoing token (if needed)
```

Comment:
This comment is confusing. We *always* get the outgoing token (and always need it) here.